pandora-server-directory/src/networking/manager_client.ts

```diff
@@ -903,3 +934,38 @@ function WithConstantTime<TParams extends unknown[], TReturn extends { /** only
 		return result[1];
 	};
 }
+
+const LoginManager = new class LoginManager {
+	private invalidAttempts: { readonly timestamp: number; }[] = [];
+
+	public loginFailed(): void {
+		const now = Date.now();
+		this.invalidAttempts.push({ timestamp: now });
+		this.invalidAttempts = this.invalidAttempts
+			.slice(-ENV.LOGIN_ATTEMPT_LIMIT)
+			.filter((attempt) => (attempt.timestamp + ENV.LOGIN_ATTEMPT_WINDOW) > now);
+	}
+
+	public async testOptionalCaptcha(data: SecondFactorData | undefined): Promise<null | SecondFactorResponse> {
+		if (!this.isCaptchaRequired()) {
+			return null;
+		}
+		const types: SecondFactorType[] = ['captcha'];
+		if (data == null) {
+			return { result: 'secondFactorRequired', types };
+		}
+		if (data.captcha == null) {
+			this.loginFailed();
```

Comment:
I don't think missing captcha should increase the counter - a valid login can trigger this easily if the client doesn't yet know to include it, which would mean that if this mechanism is triggered, it will keep stacking up the failed logins even if only valid clients continue logging in.

Right now it doesn't matter much as this only triggers if there is _some_ second factor and captcha is the only one, but if we ever introduce another second factor (like TOTP authentication) this might be problem.
I think that second factor being `undefined` and `{}` should be treated the same.